devdoc/waiter.md

```diff
@@ -0,0 +1,231 @@
+    MOCKABLE_FUNCTION(, WAITER_RESULT, waiter_notify, THANDLE(WAITER), waiter, THANDLE(RC_PTR), data, NOTIFY_COMPLETE_CALLBACK, notify_complete_callback, void*, context, THANDLE(ASYNC_OP)*, out_op);
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+```
+
+### waiter_create
+```c
+    MOCKABLE_FUNCTION(, WAITER_HANDLE, waiter_create)
+```
+
+`waiter_create` creates the waiter and returns it.
+
+**SRS_WAITER_43_001: [** `waiter_create` shall call `THANDLE_MALLOC` with `dispose` as `waiter_dispose`. **]**
+
+**SRS_WAITER_43_002: [** If there are any failures, `waiter_create` shall fail and return `NULL`. **]**
+
+### waiter_dispose
+```c
+    void waiter_dispose(WAITER* waiter)
+```
+
+**SRS_WAITER_43_003: [** `waiter_dispose` disposes the given `waiter`. **]**
+
+**SRS_WAITER_43_004: [** `waiter_dispose` shall obtain the current state of the waiter by calling `interlocked_and`. **]**
+
+**SRS_WAITER_43_005: [** If the `REGISTERED` bit is set, `waiter_dispose` shall call the `notification_callback` given to `waiter_register_notification` with `context` as the `context` given to `waiter_register_notification`, `data` as `NULL` and `result` as `WAITER_RESULT_ABANDONED`. **]**
+
+**SRS_WAITER_43_006: [** If the `NOTIFIED` bit is set, `waiter_dispose` shall: **]**
+
+- **SRS_WAITER_43_044: [** call the `notify_complete_callback` given to `waiter_notify` with `context` as the `context` given to `waiter_notify`, and `result` as `WAITER_RESULT_ABANDONED`. **]**
+
+- **SRS_WAITER_43_045: [** release the reference to `data` stored in `waiter` by calling `THANDLE_ASSIGN`. **]**
+
+### waiter_register_notification
+```c
+    MOCKABLE_FUNCTION(, WAITER_RESULT, waiter_register_notification, THANDLE(WAITER), waiter, NOTIFICATION_CALLBACK, notification_callback, void*, context, THANDLE(ASYNC_OP)*, out_op);
+```
+
+`waiter_register_notification` registers the given `notification_callback` to be called when `waiter_notify` is called.
+
+**SRS_WAITER_43_007: [** If `waiter` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_008: [** If `notification_callback` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_009: [** If `out_op` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_010: [** `waiter_register_notification` shall call `interlocked_or` to set the `REGISTERED` bit and obtain the current state of the `waiter`. **]**
+
+**SRS_WAITER_43_011: [** If the `REGISTERED` bit was already set in the current state, `waiter_register_notification` shall fail and return `WAITER_RESULT_REFUSED`. **]**
+
+**SRS_WAITER_43_012: [** If the `NOTIFIED` bit is set, `waiter_register_notification` shall : **]**
+
+- **SRS_WAITER_43_042: [** call `interlocked_exchange` to increment the `EPOCH` and unset the `NOTIFIED` bit and `REGISTERED` bit. **]**
```

Comment:
>interlocked_exchange [](http://example.com/codeflow?start=33&length=20)

This blind interlocked exchange here tells me that this is a race condition waiting to happen.
You can only do an interlocked_exchange in the case that you know 100% that there is no way that this value will be changed by other threads. Otherwise, you will overwrite what the other threads are changing.